### PR TITLE
feat(server): add journal sync endpoint

### LINKED ISSUE #155: Journal
Introduce a journal synchronization API so clients can fetch only what changed while they were offline, including deletions.

### Journal = deletions only
- New and updated items are queried directly from `SpaceItems` (`WHERE SharedAt >= checkpoint`)
- Only deletions need journaling because they disappear from the main table

### Data Model

**`DeletedItem` table**

| Column | Type | Notes |
|--------|------|-------|
| `ItemId` | `Guid` (PK) | Same as the deleted `SpaceItem.Id` |
| `SpaceId` | `Guid` (FK) | For filtering and cascade delete |
| `DeletedAt` | `DateTime` (UTC) | Timestamp of deletion |

Index: `(SpaceId, DeletedAt)`

**`Space` changes**
- `JournalPrunedBefore` (` inclusive watermark updated when old tombstones are prunedDateTime?`) 

**`SpaceMember` changes**
- `LastSyncAt` (` stored journal checkpoint / last acknowledged position for that clientDateTime?`) 

### API

```http
GET /v1/spaces/{spaceId}/journal
POST /v1/spaces/{spaceId}/journal/checkpoint
DELETE /v1/spaces/{spaceId}/journal/checkpoint
```

#### `GET /journal`
Returns the next batch of changes using the member's stored checkpoint.

**Response**
```json
{
  "fullSyncRequired": false,
  "checkpoint": "2026-04-02T08:00:00Z",
  "addedOrUpdated": [],
  "deleted": []
}
```

Behavior:
- Uses `SpaceMember.LastSyncAt` as the lower bound; `null` behaves like a full sync baseline
- Returns a server-generated `checkpoint` that the client must acknowledge after it has applied the response locally
- `fullSyncRequired: true` when the stored checkpoint is older than or equal to `JournalPrunedBefore`
- When `fullSyncRequired` is `true`, `deleted` is empty and the client should rebuild its local state from the full item list

#### `POST /journal/checkpoint`
Acknowledges that the client successfully processed a previously returned journal response.

**Request**
```json
{
  "checkpoint": "2026-04-02T08:00:00Z"
}
```

Behavior:
- Advances the member's stored checkpoint only after successful local processing
- Makes crash recovery safe: if the client dies before the `POST`, the server checkpoint does not advance

#### `DELETE /journal/checkpoint`
Clears the stored checkpoint for that member and forces the next `GET /journal` to behave like a full sync.

### Enrollment / opt-in
- Journal writes are only maintained for spaces with at least one member enrolled in journaling
- The first `GET /journal` enrolls the member for journaling, but the checkpoint only advances on `POST /journal/checkpoint`
- `DELETE /journal/checkpoint` resets the member back to "needs full sync"

### Retention Policy
- Count-based per space, configurable via `Journal:MaxEntriesPerSpace`
- Default is **1000** entries per space
- When the limit is exceeded, the oldest entries are pruned and `JournalPrunedBefore` is updated to the max pruned `DeletedAt`

### Scope
- **Server**: `DeletedItem` entity, migration, journal endpoints, pruning logic, write-on-delete
- **CLI**: consume the journal for incremental sync and deletion reconciliation (`#132`)
- **Web client**: not required for v1 because it has no local offline mirror to reconcile

Related to #132  
Related to #157

### PR DESCRIPTION
**Pull request by maraf**
Adds a crash-safe journal synchronization API so offline clients can incrementally reconcile deletions without re-fetching the full space state every time.

### Data model
- new `DeletedItem` tombstones (`ItemId`, `SpaceId`, `DeletedAt`)
- `Space.JournalPrunedBefore` watermark for journal retention gaps
- `SpaceMember.LastSyncAt` stores the member's last acknowledged checkpoint

### Journal API
- `GET /v1/spaces/{spaceId}/journal`
  - uses the stored member checkpoint
  - returns `{ fullSyncRequired, checkpoint, addedOrUpdated, deleted }`
- `POST /v1/spaces/{spaceId}/journal/checkpoint`
  - advances the member checkpoint after the client successfully applies the response
- `DELETE /v1/spaces/{spaceId}/journal/checkpoint`
  - clears the stored checkpoint and forces the next journal read to behave like a full sync

### Write path
- delete and move flows write tombstones in the same save operation as the item removal
- journal pruning is count-based per space via `Journal:MaxEntriesPerSpace` (default `1000` in `appsettings.json`)
- tombstones are only maintained for spaces with at least one enrolled journaling member

### Tests
- covers checkpoint replay-until-ack behavior, reset semantics, pruning, auth, full-sync boundaries, and delete/move integration
- current validation: `238` server tests and `54` CLI tests passing

Closes #155
Related to #132, #157

### CODE REVIEW

**Comment by maraf** [comment]
Replied to both copilot reviewer comments in the latest push (61e3d97):

**Concurrency in  SQLite is single-writer so concurrent prunes cannot race at the DB level. Noted for future-proofing if we move to Postgres.pruning** 

**`since` DateTime  Fixed: `Unspecified` DateTimeKind is now treated as UTC via `DateTime.SpecifyKind`, avoiding local-time shifts from `ToUniversalTime()`.handling**

File: src/SharedSpaces.Server/Features/Journal/JournalOptions.cs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+namespace SharedSpaces.Server.Features.Journal;
+
+public sealed class JournalOptions
+{
+    public int? MaxEntriesPerSpace { get; set; }
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/SharedSpaces.Server/Features/Journal/JournalOptions.cs:5]: This default should be in the `appsettings.json`. Do not add the default to the class itself

File: src/SharedSpaces.Server/Features/Items/ItemEndpoints.cs [hunk 2]
```diff
@@ -685,6 +710,21 @@ private static async Task<IResult> TransferItemCrossServer(
...
+                    {
+                        db.DeletedItems.Add(new DeletedItem
+                        {
+                            ItemId = sourceItemTracked.Id,
+                            SpaceId = sourceItemTracked.SpaceId,
+                            DeletedAt = DateTime.UtcNow
+                        });
+                    }
+
                     await db.SaveChangesAsync(cancellationToken);
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/SharedSpaces.Server/Features/Items/ItemEndpoints.cs:718]: added `PruneDeletedItemsAsync` call after `SaveChangesAsync` in the cross-server move path, guarded behind `hasJournalSubscribers`.

File: src/SharedSpaces.Server/Features/Items/ItemEndpoints.cs [hunk 3]
```diff
@@ -877,6 +918,20 @@ private static async Task<IResult> TransferItemSameServer(
...
+                        .AnyAsync(m => m.SpaceId == spaceId && m.LastSyncAt != null, cancellationToken);
+
+                    if (hasJournalSubscribers)
+                    {
+                        db.DeletedItems.Add(new DeletedItem
+                        {
+                            ItemId = sourceItemTracked.Id,
+                            SpaceId = sourceItemTracked.SpaceId,
+                            DeletedAt = DateTime.UtcNow
+                        });
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/SharedSpaces.Server/Features/Items/ItemEndpoints.cs]: Fixed via the shared `UpsertDeletedItemAsync` helper.

File: src/SharedSpaces.Server/Features/Journal/JournalOptions.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (maraf) [reply src/SharedSpaces.Server/Features/Journal/JournalOptions.cs:5]: Per owner feedback: default added to `appsettings.json` (`Journal:MaxEntriesPerSpace: 1000`) rather than in the options class itself.

File: src/SharedSpaces.Server/Features/Items/ItemEndpoints.cs [hunk 4]
```diff
@@ -685,6 +710,21 @@ private static async Task<IResult> TransferItemCrossServer(
...
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/SharedSpaces.Server/Features/Items/ItemEndpoints.cs]: Fixed via the shared `UpsertDeletedItemAsync`  checks for existing tombstone before adding.

File: src/SharedSpaces.Server/Features/Items/ItemEndpoints.cs [hunk 5]
```diff
@@ -413,8 +415,29 @@ private static async Task<IResult> DeleteItem(
         var isFile = string.Equals(item.ContentType, "file", StringComparison.OrdinalIgnoreCase);
 
         db.SpaceItems.Remove(item);
+
+        // Only write journal tombstone if at least one member has opted into journaling
+        var hasJournalSubscribers = await db.SpaceMembers
+            .AnyAsync(m => m.SpaceId == spaceId && m.LastSyncAt != null, cancellationToken);
+
+        if (hasJournalSubscribers)
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/SharedSpaces.Server/Features/Items/ItemEndpoints.cs:423]: This is intentional per owner review  most spaces will never have journaling members, so gating avoids writing tombstones that no one will ever read. The PR description will be updated to reflect this design.

File: src/SharedSpaces.Server/Features/Items/ItemEndpoints.cs [hunk 6]
```diff
@@ -413,8 +415,29 @@ private static async Task<IResult> DeleteItem(
...
+            .AnyAsync(m => m.SpaceId == spaceId && m.LastSyncAt != null, cancellationToken);
+
+        if (hasJournalSubscribers)
+        {
+            db.DeletedItems.Add(new DeletedItem
+            {
+                ItemId = item.Id,
+                SpaceId = item.SpaceId,
+                DeletedAt = DateTime.UtcNow
+            });
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/SharedSpaces.Server/Features/Items/ItemEndpoints.cs]: extracted `UpsertDeletedItemAsync` helper that checks for an existing tombstone and updates it (or creates new). Applied to all 3 paths: DeleteItem, TransferItemCrossServer, TransferItemSameServer.

File: src/SharedSpaces.Server/Features/Items/ItemEndpoints.cs [hunk 7]
```diff
@@ -919,6 +974,9 @@ private static async Task<IResult> TransferItemSameServer(
                         // Best-effort cleanup
                     }
                 }
+
+                // Prune old journal entries if configured
+                await PruneDeletedItemsAsync(db, spaceId, journalOptions.Value, cancellationToken);
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/SharedSpaces.Server/Features/Items/ItemEndpoints.cs]: pruning is now guarded behind `if (hasJournalSubscribers)`, consistent with DeleteItem and cross-server paths.

File: src/SharedSpaces.Server/Features/Journal/JournalEndpoints.cs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+        if (space is null)
+        {
+            return Results.NotFound(new { Error = "Space not found" });
+        }
+
+        var sinceUtc = since.HasValue
+            ? DateTime.SpecifyKind(since.Value, DateTimeKind.Utc)
+            : DateTime.MinValue;
+        var fullSyncRequired = space.JournalPrunedBefore.HasValue && sinceUtc < space.JournalPrunedBefore.Value;
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/SharedSpaces.Server/Features/Journal/JournalEndpoints.cs]: changed to `sinceUtc <= space.JournalPrunedBefore.Value`. Added test `GetJournal_ReturnsFullSyncRequired_WhenSinceEqualsWatermark` to cover the boundary case.

File: src/SharedSpaces.Server/Features/Journal/JournalEndpoints.cs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        if (authorizationResult is not null)
+        {
+            return authorizationResult;
+        }
+
+        // Treat Unspecified as UTC; reject ambiguous local-time offsets
+        if (since.HasValue && since.Value.Kind == DateTimeKind.Local)
+        {
+            return Results.BadRequest(new { Error = "The 'since' parameter must be in UTC (use Z suffix or no offset)." });
+        }
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/SharedSpaces.Server/Features/Journal/JournalEndpoints.cs]: switched `since` parameter from `DateTime?` to `DateTimeOffset?`. This naturally handles `+00:00` offsets and any explicit timezone. We convert to UTC via `.UtcDateTime` and no longer need the DateTimeKind check.

File: src/SharedSpaces.Server/Features/Journal/JournalOptions.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (maraf) [reply src/SharedSpaces.Server/Features/Journal/JournalOptions.cs:5]: added `Journal:MaxEntriesPerSpace: 1000` to `appsettings.json`. The `JournalOptions` class stays with `int? MaxEntriesPerSpace` (no default), so removing the config disables pruning entirely.

File: src/SharedSpaces.Server/Features/Journal/JournalEndpoints.cs [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+                item.SharedAt))
+            .ToArrayAsync(cancellationToken);
+
+        Guid[] deleted;
+        if (fullSyncRequired)
+        {
+            deleted = [];
+        }
+        else
+        {
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/SharedSpaces.Server/Features/Journal/JournalEndpoints.cs]: Fixed in `cf50a09`: the handler now short-circuits when `fullSyncRequired` is true and returns empty `addedOrUpdated`/`deleted` arrays instead of running the incremental queries.

File: src/SharedSpaces.Server/Features/Journal/JournalEndpoints.cs [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        }
+
+        if (request.Checkpoint.Kind == DateTimeKind.Local)
+        {
+            return Results.BadRequest(new { Error = "The 'checkpoint' value must be in UTC." });
+        }
+
+        var checkpointUtc = request.Checkpoint.Kind == DateTimeKind.Unspecified
+            ? DateTime.SpecifyKind(request.Checkpoint, DateTimeKind.Utc)
+            : request.Checkpoint.ToUniversalTime();
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/SharedSpaces.Server/Features/Journal/JournalEndpoints.cs]: Fixed in `cf50a09`: `JournalCheckpointRequest.Checkpoint` now uses `DateTimeOffset`, and the server normalizes it via `.UtcDateTime`, so the ack payload is offset-aware and the contract is no longer ambiguous.

File: src/SharedSpaces.Server/Features/Journal/JournalEndpoints.cs [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+public static class JournalEndpoints
+{
+    public static IEndpointRouteBuilder MapJournalEndpoints(this IEndpointRouteBuilder app)
+    {
+        var group = app.MapGroup("/v1/spaces/{spaceId:guid}/journal")
+            .RequireAuthorization();
+
+        group.MapGet("/", GetJournal);
+        group.MapPost("/checkpoint", UpdateCheckpoint);
+        group.MapDelete("/checkpoint", ResetCheckpoint);
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/SharedSpaces.Server/Features/Journal/JournalEndpoints.cs:18]: Updated the PR description to match the implemented checkpoint-based contract: `GET /journal`, `POST /journal/checkpoint`, and `DELETE /journal/checkpoint`. The issue text was updated as well.
